Port text-properties-at

Bira opened the pull request:
This is my initial attempt at porting the text-properties-at function. 

The code as it is right now causes a failure in `test/lisp/emacs-lisp/ert-x-tests.el`, in the tests for `propertized-string`:

```
Test ert-propertized-string condition:
    (args-out-of-range
     #("abcd" 1 2
       (a b)
       2 4
       (c t))
     4 5)
   FAILED  2/9  ert-propertized-string
```

I'm pretty sure I'm doing something wrong in the call to validate_interval_range, but unfortunately my knowledge isn't great enough to figure out what the right way to call it would be. Any help on that would be appreciated :).

Also, is there a way to run a single lisp test file?

Bira committed:
Initial attempt at porting text-properties-at

Bira committed:
Fix comment style

brotzeit commented on rust_src/src/textprop.rs:
```diff
@@ -28,7 +29,39 @@ pub fn get_char_property(position: EmacsInt, prop: LispObject, object: LispObjec
+/// If OBJECT is a string, POSITION is a 0-based index into it.
+/// If POSITION is at the end of OBJECT, the value is nil.
+#[lisp_fn(min = "1")]
+pub fn text_properties_at(mut position: LispObject, mut object: LispObject) -> LispObject {
```
I think you can make position a `LispNumber`.

Bira replied:
I managed it but ended up breaking a lot more tests than before :). I'm sure it will be possible once validate_interval_range is ported.

Bira committed:
Another attempt at making this test pass, with no success

shaleh commented on rust_src/src/textprop.rs:
```diff
@@ -28,7 +29,39 @@ pub fn get_char_property(position: EmacsInt, prop: LispObject, object: LispObjec
+    let i: *mut Lisp_Interval =
+        unsafe { validate_interval_range(object, position_pointer, position_pointer, false) };
+
+    if ptr::eq(i, ptr::null_mut()) {
```
```suggestion
    if i.is_null() {
```

`ptr` provides `is_null`.

Bira committed:
Rust raw pointers provide is_null()

Bira committed:
Can't pass the same pointer twice to validate_interval_range

Bira commented:
I finally found out what was causing the text failure I was getting here. I was totally wrong about the workings of the LENGTH macro for intervals. I solved this by adding a minimal LispIntervalRef implementation with only the things I needed for this specific case. 

Is there anything else I need to add to that before it can be considered a true "minimal" implementation ready for future expansion?

shaleh commented on rust_src/src/textprop.rs:
```diff
@@ -30,7 +30,42 @@ pub fn get_char_property(position: LispNumber, prop: LispObject, object: LispObj
+        object = ThreadState::current_buffer().into();
+    }
+
+    let i: LispIntervalRef = LispIntervalRef::new(unsafe {
```
`validate_interval_range` is used as a constructor in the C code except in one place where it is only used for the validation. A more appropriate design for `LispIntervalRef` I think would be to add a `new_from_position()` factory method which takes the object and the position (only once). In this use case the final option is always false in the C code.

`let i: LispIntervalRef::new_from_position(object, position);`

This does however lose the "validated" which is likely important for the reader. Should the name be `new_validated_from_position`? `new_validated_position`? `new_position_with_validation`?

Bira replied:
I have it as `from_valid_range` locally, with the validation itself separated into another function that's called internally. It seems to work in that the tests still pass, but I'm not 100% clear on what the XSETFASTINT macro does on the C version. It's being called to assign the value of `*begin` to itself, but I couldn't quite parse the bitwise operations it does. Is that where they increment these numbers by 1?

shaleh replied:
For Rust we can essentially read `XSETFASTINT` as `let x: LispObject = LispObject::from(some_number);`.

Bira commented:
I've submitted my implementation of `LispIntervalRef::from_valid_range()` for review. It's _almost_ a port of `validate_interval_range` - should it be approved, the actual port could consist of wrapping this method in an extern C function.

Bira commented:
To clarify my comment from earlier, I didn't want to port `validate_interval_range` itself as part of this PR, but to use the new Rust method for the port of `text_properties_at`. I'm not quite sure the method would work for a general port as-is, since it's missing some of the pointer-wrangling of the C version, and I would rather investigate that in a separate PR. 

Of course, if you think it's better to do the full port in this branch I will work on it.

shaleh commented:
If that is the case, perhaps the right thing to do is have the method you created simply call the C function and then in a follow in PR you can replace the C call with the fully functioning Rust code.

Bira committed:
Add minimal interval length logic.

This code stands in for the TOTAL_LENGTH and LENGTH macros defined in
intervals.h. The most correct approach here would be to implement a
LispIntervalRef type that properly encapsulated this logic as methods,
but this is a bit outside the scope of this specific branch. I'll do it
separately.

Bira commented:
I've removed the LispIntervalRef changes to a separate branch, and implemented a private function with the absolute minimum logic I need to make text-properties-at work correctly. I would be glad to proceed with the LispIntervalRef work in a separate PR.

Bira commented:
Looks like that full port of validate_interval_range has enough "hidden" issues with it that it's a bit beyond my ability to fix properly. Still, does the code in this branch look good enough to merge?

shaleh commented:
@Bira do you have time to refresh it?

Bira commented:
I'll try to do that this week!

Bira committed:
Fix imports after rebasing on latest master

Bira commented:
I've refreshed it!
